File: leetcode/cpp/29_divide_two_integers.cpp
#include <limits.h>
#include <stdint.h>

class Solution {

  public:
    int divide(int dividend, int divisor)
    {
      if (divisor == 0) return INT_MAX;
      bool flag = true;
      if ((dividend > 0 && divisor < 0) || (dividend < 0 && divisor > 0)) {
        flag = false;
      }
      int64_t n = dividend, m = divisor;
      n = n < 0 ? -n : n;
      m = m < 0 ? -m : m;
      int64_t c = 0, t, temp;
      while (n >= m) {
        t = 1, temp = m;
        while ((temp<<1) <= n) {
          temp <<= 1;
          t <<= 1;
        }
        c += t;
        if (flag && c > INT_MAX) return INT_MAX;
        if (!flag && c - INT_MAX > 1) return INT_MAX;
        n -= temp;
      }
      return flag ? c : -c;
    }
};

File: leetcode/go/src/leetcode/29_divide_two_integers.go
package leetcode

func divide(dividend int, divisor int) int {
	// signed flag
	f := -1
	if dividend > 0 && divisor > 0 || dividend < 0 && divisor < 0 {
		f = 1
	}

	var (
		res int64
		i   int64 = 1
		m         = int64(divisor)
		d         = int64(dividend)
	)
	if divisor < 0 {
		m = -m
	}
	if dividend < 0 {
		d = -d
	}

	// minus by exponentially
	for d >= m {
		d = d - m
		res = res + i
		i = i << 1
		m = m << 1
	}

	// minus left part
	m = int64(divisor)
	if m < 0 {
		m = -m
	}
	for d >= m {
		d = d - m
		res = res + 1
	}

	// check overflow
	if f == 1 && res > (1<<31-1) {
		return (1<<31 - 1)
	}

	if f == -1 && res > (1<<31) {
		return -(1 << 31)
	}

	return f * int(res)
}
